File: LogicOverSyntaxCodes/Searching/JavaCodes/Leetcode410.java
class Solution {
    public static boolean canISplitAtMaxK(int[] nums,int k,int maxSum){
        int tempK=1,sum=0;

        for (int i = 0; i < nums.length; i++) {
            if(sum+nums[i]<=maxSum){
                sum+=nums[i];
            }
            else{
                tempK++;
                sum=nums[i];
            }
        }
        return tempK<=k;
    }

    public static int splitArray(int[] nums, int k) {
        int max=0,sum=0;
        for (int i = 0; i < nums.length; i++) {
            max=Math.max(max,nums[i]);
            sum+=nums[i];
        }
        int start=max,end=sum;

        while(start<end){
            int mid=start+(end-start)/2;
            if(canISplitAtMaxK(nums,k,mid)){
                end=mid;
            }
            else{
                start=mid+1;
            }
        }
        return end;
    }
}

File: Searching/PythonCodes/Leetcode410.py
class Solution(object):
    def canISplitAtMaxK(self, nums, k, maxSum):
        tempK, total = 1, 0
        
        for num in nums:
            if total + num <= maxSum:
                total += num
            else:
                tempK += 1
                total = num
        
        return tempK <= k
    
    def splitArray(self, nums, k):
        max_num, total_sum = max(nums), sum(nums)
        start, end = max_num, total_sum
        
        while start < end:
            mid = start + (end - start) // 2
            if self.canISplitAtMaxK(nums, k, mid):
                end = mid
            else:
                start = mid + 1
        
        return end